bool valid
uint32 preID
uint32 candidateID
